Sim1sbind

taustinn11 opened the pull request:
Use this pull request - fixed .gitignore

Commit:
Added sandbox.rmd to Rbuildignore to keep out of the package

started with sandbox
working on one-site total binding Eq
Having to troubleshoot log transformation

Commit:
Can coerce graphing of log = T values by using loess method with y ~ x formula.

However, going to need to try different values as start arguments and see if we can get nlsLM to work

Commit:
Added comments.

Importantly, the loess method starts to look very poor as you increase the cv (obviously). This cannot be used.

Commit:
Use stat_function call to bypass nlsLM error. Tested with higher values and higher variance and seems to still plot well

Generated a .R file for it using use_r()

Commit:
Updated sim1sbind to plot a stat_function based on parameters gathered from nls call in the log-linear -> coefficients hack

Vast improvement because now stat_function call plots a line that responds to the variation in simulated points

Commit:
quick fix

Commit:
Added sandbox.rmd to .gitignore - keeping for local to test

Commit:
.gitignore fix

TJMurphy commented on sandbox.rmd:
```diff
@@ -384,3 +384,56 @@ sim1sbind <- function(x, bmax, kd, cv, reps, log=F) {
+      log10(x)} else {
+        x}, y)) +
+    ggplot2::geom_point(size=2) +
+    ggplot2::labs(title="model: y=bmax*x/(x+kd)") +
```
add to labs: x="x"

TJMurphy replied:
it would be nice for when log=T that xlab = "logx" and title="y=bmax*logx/(logx+logkd)

taustinn11 replied:
Done! (Took longer than the other ones -- see my most recent push)

TJMurphy commented on sandbox.rmd:
```diff
@@ -384,3 +384,56 @@ sim1sbind <- function(x, bmax, kd, cv, reps, log=F) {
 
+Let's update this:
+
+right now, the stat_function call is just plotting an exact line for the input values of bmax and kd. It is not responding to the variability of the chosen points. 
```
maybe it would be better to have two functions? one for linear x entry, another for log x entry. That way you wouldn't need stat_function, and could draw the line with geom_smooth

taustinn11 replied:
See my below comment

Also, all final changes I put in the <function_name>.R. You can basically ignore sandbox.rmd as far as commentary goes. It just shows the successive iterations until I get to the final function which I take into the .R file

TJMurphy commented on R/sim1sbind.R:
```diff
@@ -0,0 +1,58 @@
+#' Simulate nonlinear ligand-receptor binding with one binding site
```
"Simulate one-site ligand-receptor binding"

taustinn11 replied:
Done!

TJMurphy commented on R/sim1sbind.R:
```diff
@@ -0,0 +1,58 @@
+#'
+#' A sandbox to simulate and visualize random normal heteroscedastic response data. Variances enlarge with the value of y predicted by the model using a constant coefficeint of variation (cv). The data generating formula is derived from the one-site total binding model: y=Bmax*x/(x+kd). Failure errors in the plot fitting subfunction will occasionally happen due to the random data. These are more  ...
+#'
+#' @param x a vector of non-exponential linear scale values, usually representing dose or concentration, but can represent any stimulus.
```
strike "but can represent any stimulus"  because this is binding, not stimulation

taustinn11 replied:
Done!

TJMurphy commented on R/sim1sbind.R:
```diff
@@ -0,0 +1,58 @@
+#' @param kd the value of x that yields y/Bmax = 0.5 (the equilibrium binding constant)
+#' @param cv the coefficient of variation for y replicates.
+#' @param reps an integer value for number of replicates
+#' @param log logical value. Default is FALSE. If TRUE, linear x values are transformed using a log10 function for plotting. Only for visual aesthetic.
```
we might want to omit this option, create a function for log x entry

taustinn11 replied:
I really don't think we need to do that. I think it's more powerful to keep it wrapped into one function.

The stat_function() call bypasses plotting errors when using geom_smooth(method = "nlsLM") like "singular gradient matrix at start" or "max iterations exceeds 50" or "step halving factor reduced below minimum" (I paraphrased the errors from memory).

Essentially, I use nls to generate estimates for Bmax and Kd (called bmax.var and kd.var). That is what is done in the for loop from lines 37 - 46. Then, stat_function() will plot the model equation and use bmax.var and kd.var and use geom_smooth to curve the line. I've tested it with several different values for x, bmax, kd, and cv, and the stat_function() call modifies the curve accordingly.

TJMurphy commented on sandbox.rmd:
```diff
@@ -227,3 +236,59 @@ simbindone(x, bmax, kd, 0.01, 5, log = T)
+  
+  for (i in 1) {
+    if (log) {
+      model <- stats::lm(log(y) ~ log(x), data=values)
```
We log x only, to flip the x-axis so the dose response is S shaped rather than hyperbolic. We don't log values of y.

taustinn11 replied:
This fix is only to generate values (stored in model) that are then used to be the start values for the nls call in the next line. The final graph does not have log y values

Commit:
Altered some wording of params (as per TJ request)

Commit:
Changed title

Clarified labels in if (log) conditionals

Commit:
Reformatted to have multiple if conditionals - should fix error about adding ggprotos together

Commit:
Stored ggplot calls (with conditionals) into objects and then printed them at the end

Allows for better labeling of the axes as far as I can figure out right now

Commit:
Found a workaround for programming the labels and geom_smooth calls for if (log). It now returns the ggplot call without needing to add objects together or use a print command

Commit:
Check() after ensuring function works appropriately

taustinn11 commented:
@TJMurphy See my most recent commit (184e8a6). It has what we need.
Also, see my comment about whether we should separate linear vs. log

Commit:
modified @param x to be more accurate

added example using exponential values for x

changed stats::nls to minpack.lm::nlsLM for log = T parameter estimation

Commit:
quick check()

modified logdose in @examples to not be too long (gave be a note with check())

TJMurphy committed:
Update sim1sbind.R
